### ISSUE TITLE
One-line configuration of the kubernetes client

### ISSUE THREAD

**Issue by afbjorklund** [lima-vm/lima#1300]
### Description

This is a re-iteration of an earlier feature request, that was deemed redundant:

* https://github.com/lima-vm/lima/issues/514

I would like to make the config more similar to the config of Docker / Podman.

This by letting lima do the steps of creating the config directory and the file...

----

That way, we don't have to carry the shell scripts and config locations around.

Making it more similar, to how we tunnel the docker.sock and podman.sock ?

`export DOCKER_HOST=$(limactl list docker --format 'unix://{{.Dir}}/sock/docker.sock')`

`export CONTAINER_HOST=$(limactl list podman --format 'unix://{{.Dir}}/sock/podman.sock')`

So the request is to make ConfigFile into a configuration item, like PortForward:

```diff
diff --git a/examples/k3s.yaml b/examples/k3s.yaml
index c3d0769..438b018 100644
--- a/examples/k3s.yaml
+++ b/examples/k3s.yaml
@@ -5,8 +5,7 @@
 # It can be accessed from the host by exporting the kubeconfig file;
 # the ports are already forwarded automatically by lima:
 #
-# $ export KUBECONFIG=$PWD/kubeconfig.yaml
-# $ limactl shell k3s sudo cat /etc/rancher/k3s/k3s.yaml >$KUBECONFIG
+# $ export KUBECONFIG=$(limactl list k3s --format 'unix://{{.Dir}}/conf/kubeconfig.yaml')
 # $ kubectl get no
 # NAME       STATUS   ROLES                  AGE   VERSION
 # lima-k3s   Ready    control-plane,master   69s   v1.21.1+k3s1
@@ -54,11 +53,12 @@ probes:
     The k3s kubeconfig file has not yet been created.
     Run "limactl shell k3s sudo journalctl -u k3s" to check the log.
     If that is still empty, check the bottom of the log at "/var/log/cloud-init-output.log".
+configFiles:
+- guestFile: "/etc/rancher/k3s/k3s.yaml"
+  hostFile: "{{.Dir}}/conf/kubeconfig.yaml"
 message: |
   To run `kubectl` on the host (assumes kubectl is installed), run the following commands:
   ------
-  mkdir -p "{{.Dir}}/conf"
   export KUBECONFIG="{{.Dir}}/conf/kubeconfig.yaml"
-  limactl shell {{.Name}} sudo cat /etc/rancher/k3s/k3s.yaml >$KUBECONFIG
   kubectl ...
   ------
diff --git a/examples/k8s.yaml b/examples/k8s.yaml
index 088ce25..44f8aee 100644
--- a/examples/k8s.yaml
+++ b/examples/k8s.yaml
@@ -5,8 +5,7 @@
 # It can be accessed from the host by exporting the kubeconfig file;
 # the ports are already forwarded automatically by lima:
 #
-# $ export KUBECONFIG=$PWD/kubeconfig.yaml
-# $ limactl shell k8s sudo cat /etc/kubernetes/admin.conf >$KUBECONFIG
+# $ export KUBECONFIG=$(limactl list k8s --format 'unix://{{.Dir}}/conf/kubeconfig.yaml')
 # $ kubectl get no
 # NAME       STATUS   ROLES                  AGE   VERSION
 # lima-k8s   Ready    control-plane,master   44s   v1.22.3
@@ -155,11 +154,12 @@ probes:
       echo >&2 "kubernetes cluster is not up and running yet"
       exit 1
     fi
+configFiles:
+- guestFile: "/etc/kubernetes/admin.conf"
+  hostFile: "{{.Dir}}/conf/kubeconfig.yaml"
 message: |
   To run `kubectl` on the host (assumes kubectl is installed), run the following commands:
   ------
-  mkdir -p "{{.Dir}}/conf"
   export KUBECONFIG="{{.Dir}}/conf/kubeconfig.yaml"
-  limactl shell {{.Name}} sudo cat /etc/kubernetes/admin.conf >$KUBECONFIG
   kubectl ...
   ------
```

It just runs `cat`, over ssh. There is configuration (called "mode"), for the `sudo`.

**Comment by afbjorklund** [comment]
Kubernetes are apparently evaluating alternatives, to the `minikube start`. For the "Getting Started" tutorials, that is:

https://kubernetes.io/docs/tutorials/hello-minikube/

I would like to make `limactl start` a candidate, without having to use all of Rancher Desktop or Colima or something.

The other two candidates are running in the cloud (killercoda), or running nested in Docker with `kind` or with `k3d`.

**Comment by afbjorklund** [comment]
One _could_ also do similar hacks as the current docker.lima script, like kubectl.lima:

https://github.com/lima-vm/lima/blob/master/cmd/docker.lima

```bash
#!/bin/sh
set -eu
: "${LIMA_INSTANCE:=k8s}"
: "${KUBECTL:=kubectl}"

if [ "$(limactl ls -q "$LIMA_INSTANCE" 2>/dev/null)" != "$LIMA_INSTANCE" ]; then
  echo "instance \"$LIMA_INSTANCE\" does not exist, run \`limactl start --name=$LIMA_INSTANCE template://k8s\` to create a new instance" >&2
  exit 1
fi
KUBECTL=$(command -v "$KUBECTL" || true)
if [ -n "$KUBECTL" ]; then
  KUBECONFIG=$(limactl list "$LIMA_INSTANCE" --format '{{.Dir}}/conf/kubeconfig.yaml')
  export KUBECONFIG
  exec "$KUBECTL" "$@"
else
  export LIMA_INSTANCE
  exec lima sudo kubectl "$@"
fi
```

And we can argue if it should be `LIMA_INSTANCE=k3s KUBECTL="k3s kubectl"` - or not.

**Comment by jandubois** [comment]
> One _could_ also do similar hacks as the current docker.lima script, like kubectl.lima:

I think this is fine. We need to draw the line somewhere, but given that even apptainer has a lima wrapper, k8s definitely deserves to have one too.

> And we can argue if it should be `LIMA_INSTANCE=k3s KUBECTL="kubectl"` - or not.

Maybe

```shell
if [ -z "$LIMA_INSTANCE" ]; then
    if [ "$(limactl ls -f '{{.Status}}' k3s 2>/dev/null)" = "Running" ]; then
        LIMA_INSTANCE=k3s
    elif [ "$(limactl ls '{{.Status}}' k8s 2>/dev/null)" = "Running" ]; then
        LIMA_INSTANCE=k8s
    fi
fi
```

That way the script choses `k3s` or `k8s` if they are running, and still default to `default` if neither one are (user will have to setup Kubernetes in `default` themselves).

Maybe that is too magical, but I think it probably will do the right thing for most users most of the time, and that is the main point of bundling a helper script.

**Comment by afbjorklund** [comment]
Or just error out perhaps, like the other ones ? "Please start one of the bundled Kubernetes templates"

**Comment by afbjorklund** [comment]
With "kubectl wrappers", I meant there is a large number of projects to address the version craziness.

The client has to match the server API, within a certain amount of [version skew,](https://kubernetes.io/releases/version-skew-policy/) or it will fail to connect.

So there are projects that will look up the cluster version (somehow), and exec the right version binary...

* https://github.com/flavio/kuberlr



Even wrote some similar ones myself.

They will all be $KUBECTL in this context.

**Comment by jandubois** [comment]
> Not sure if the running check is even needed, the error message seems to be quite OK if it is not ?

I did the "Running" check in case you have both a stopped `k3s` instance and a running `k8s` instance. 😄 

> So there are projects that will look up the cluster version (somehow), and exec the right version binary...

I know:

```console
$ readlink "$(which kubectl)"
/Applications/Rancher Desktop.app/Contents/Resources/resources/darwin/bin/kubectl

$ readlink "$(readlink "$(which kubectl)")"
kuberlr
```

**Comment by afbjorklund** [comment]
The fallback is to run kubectl on the node directly, that should at least be the correct version...

Removed the templates, added a check in case you override the LIMA_INSTANCE yourself.

**Comment by afbjorklund** [comment]
Closed in PR #1302
